metrictank/src/main/java/com/expedia/metrics/metrictank/MetricTankIdFactory.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2018 Expedia Group, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.expedia.metrics.metrictank;
+
+import com.expedia.metrics.IdFactory;
+import com.expedia.metrics.MetricDefinition;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.*;
+
+import static com.expedia.metrics.metrictank.MessagePackSerializer.INTERVAL;
+import static com.expedia.metrics.metrictank.MessagePackSerializer.ORG_ID;
+
+public class MetricTankIdFactory implements IdFactory {
+    private static final char[] HEX_DIGITS = {'0','1','2','3','4','5','6','7','8','9','a','b','c','d','e','f'};
+
+    @Override
+    public String getId(MetricDefinition metric) {
+        Map<String, String> tags = new HashMap<>(metric.getTags().getKv());
+        final int orgId;
+        try {
+            orgId = Integer.parseInt(tags.remove(ORG_ID));
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("Org ID must be an int", e);
+        }
+        final String name = metric.getKey();
+        if (name == null) {
+            throw new IllegalArgumentException("Key is required by metrictank");
```

Comment:
Using "Key", "Unit", "Mtype" here may be confusing since the actual key names are "key", "unit", and "mtype". I am inclined to treat the tags as case-sensitive just because the spec talks about casing conventions for tag names and doesn't ever say "these should be case insensitive".

Reply:
If you don't like starting the sentence with a lowercase letter, consider

Tag "unit" is required by MetricTank.

Reply:
done